dali/pipeline/operators/crop/crop.h

```diff
@@ -44,32 +49,28 @@ class CropAttr {
     }
   }
 
+  /**
+   * @brief Calculate coordinate where the crop starts in pixels.
+   *
+   * TODO(klecki) this should produce (crop_[0] * W, crop_[1] * H) and is broken, (FIXME)
+   *
+   * @param spec
+   * @param ws
+   * @param imgIdx
+   * @param H
+   * @param W
+   * @return std::pair<int, int>
+   */
   std::pair<int, int> SetCropXY(const OpSpec &spec, const ArgumentWorkspace *ws,
-                                const Index dataIdx, int H, int W) {
-    DALI_ENFORCE(H >= crop_height_[dataIdx]);
-    DALI_ENFORCE(W >= crop_width_[dataIdx]);
+                                const Index dataIdx, int H, int W);
 
-    auto crop_x_norm = spec.GetArgument<float>("crop_pos_x", ws, dataIdx);
-    auto crop_y_norm = spec.GetArgument<float>("crop_pos_y", ws, dataIdx);
-
-    DALI_ENFORCE(crop_y_norm >= 0.f && crop_y_norm <= 1.f,
-                 "Crop coordinates need to be in range [0.0, 1.0]");
-    DALI_ENFORCE(crop_x_norm >= 0.f && crop_x_norm <= 1.f,
-                 "Crop coordinates need to be in range [0.0, 1.0]");
-
-    const int crop_y = crop_y_norm * (H - crop_height_[dataIdx]);
-    const int crop_x = crop_x_norm * (W - crop_width_[dataIdx]);
-
-    return std::make_pair(crop_y, crop_x);
-  }
-
-  const vector<Index> CheckShapes(const SampleWorkspace *ws) {
-    const auto &input = ws->Input<CPUBackend>(0);
-
-    DALI_ENFORCE(input.shape().size() == 3, "Expects 3-dimensional image input.");
-
-    return input.shape();
-  }
+  /**
+   * @brief Enforce that all shapes match
+   *
+   * @param ws
+   * @return const vector<Index> One matching shape for all inputs
+   */
+  virtual const vector<Index> CheckShapes(const SampleWorkspace *ws);
```

Comment:
Maybe the wording of this comment is not ideal, but the function came from existing implementation of Crop, that assumes that all inputs have the same shape. So we use DALI_ENFORCE to check it. "Enforce" in this project does never change anything, it works like an assert, so I'm not sure from where this idea of shape change is coming from.

If it's checking if the sizes are ok, and returning the safe one it does the same thing as checking if arguments are ok and if so, obtaining them, or setting them to Operator attribute.

However, I agree, that this is not perfectly clear, I will try to do something about it.

Reply:
So I looked a bit into the problem, and I don't see a easy way of getting rid of the assumption, that all inputs should be the same.

CheckShape and all the calculation of crop sizes are done in `SetupSharedSampleParams`, which is called in Operator::Run. At this point we have data_idx, and thread_idx already available, but, than we enter the loop over `input_sets_` and only then, when the RunImpl is called we get to know which input we will be handling. So this would require a bigger change in CPU Crop Op to move data setup from `SetupSharedSampleParams` to RunImpl.